### PR TITLE
Add device lock support

### PR DESCRIPTION
**Pull request by codingllama**
Add support for locking trusted devices. This includes forbidding certificate issuance while using/for a locked device, as well as blocking ongoing and future accesses for various services (SSH, DB, K8s, etc).

Device locks require the use of an enrolled device, otherwise the device itself isn't identified by Teleport. This means that device locking is only effective in scenarios where device trust is required.

Lock support is an exercise in tracking LockTargets. Most lock targets "flow" from the following methods:

* lib/srv.ComputeLockTargets (SSH connections)
* lib/services.LockTargetsFromTLSIdentity (TLS connections)
  * lib/auth.Context.LockTargets calls LockTargetsFromTLSIdentity

Locks for certificate issuance are gated by:

* lib/auth.verifyLocksForUserCerts
  * lib/auth.AugmentContextUserCertificates calls verifyLocksForUserCerts, used to add device extensions
  * lib/auth.generateUserCert calls verifyLocksForUserCerts, issues user certificates for all other scenarios

https://github.com/gravitational/teleport.e/issues/514

### CODE REVIEW

**Comment by codingllama** [comment]
Apart from protos and added tests, which are the bulk of the PR, this is a rather small production-code change, so it seemed reasonable to mail it all together.

Manually tested cert issuance, SSH, DB and K8s access, including:

- Session MFA on/off
- Establishing new connections
- Cutting off existing connections

Seems to be working fine. Notably, this would block commands as `tsh app login`, which usually are not affected by device trust, but that seems to be WAI to me.

**Comment by codingllama** [comment]
@smallinsky, would you mind taking a quick look to see if I missed adding lock targets or something else? No need for a full review.

If this looks reasonable I'll bookmark it for future reference. :)

File: api/types/lock.go [hunk 1]
```diff
@@ -201,28 +201,14 @@ func (t LockTarget) Match(lock Lock) bool {
...
-	}
-	return true
+	return (t.User == "" || lockTarget.User == t.User) &&
+		(t.Role == "" || lockTarget.Role == t.Role) &&
+		(t.Login == "" || lockTarget.Login == t.Login) &&
+		(t.Node == "" || lockTarget.Node == t.Node) &&
+		(t.MFADevice == "" || lockTarget.MFADevice == t.MFADevice) &&
+		(t.WindowsDesktop == "" || lockTarget.WindowsDesktop == t.WindowsDesktop) &&
+		(t.AccessRequest == "" || lockTarget.AccessRequest == t.AccessRequest) &&
+		(t.Device == "" || lockTarget.Device == t.Device)
```
Comment (tobiaszheller) [inline api/types/lock.go:211]: i am not sure if this is more readable but not strong opinions here.

**Review by tobiaszheller — APPROVED** [review]

File: api/types/lock.go (hunk 1 above)
Comment (codingllama) [reply api/types/lock.go:211]: 👍

File: lib/auth/auth.go [hunk 2]
```diff
@@ -1521,6 +1521,7 @@ func (a *Server) AugmentContextUserCertificates(
 		username:             identity.Username,
 		mfaVerified:          identity.MFAVerified,
 		activeAccessRequests: identity.ActiveRequests,
+		deviceID:             opts.DeviceExtensions.DeviceID, // Check lock against requested device.
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/auth.go:1524]: DeviceExtensions is not a pointer, so this is safe. We also validate `opts` at the start of the method.

File: lib/srv/ctx_test.go [hunk 3]
```diff
@@ -219,3 +220,67 @@ func TestIdentityContext_GetUserMetadata(t *testing.T) {
...
+				},
+			},
+			AccessChecker: &fixedRolesChecker{
+				roleNames: mappedRoles,
+			},
+			UnmappedRoles:  unmappedRoles,
+			ActiveRequests: accessRequests,
+		}
+
+		// Test.
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/srv/ctx_test.go:253]: I tend to keep these to distinguish set from the method being tested, although in this case it's quite clear. Dropped.

File: api/proto/teleport/legacy/types/types.proto [hunk 4]
```diff
@@ -4141,6 +4141,10 @@ message LockTarget {
 
   // AccessRequest specifies the UUID of an access request.
   string AccessRequest = 7 [(gogoproto.jsontag) = "access_request,omitempty"];
+
+  // Device is the device ID of a trusted device.
+  // Requires Teleport Enterprise.
+  string Device = 8 [(gogoproto.jsontag) = "device,omitempty"];
```
[the remark this answers is not in the document]
Comment (codingllama) [reply api/proto/teleport/legacy/types/types.proto:4147]: I'm following the RFD, so I'm a bit hesitant to change it now, even though I see your point. The `ID` suffix is also absent from Node, MFADevice, etc, so I think we tend to go with more user-directed names here.

https://github.com/gravitational/teleport.e/blob/master/rfd/0001e-device-trust.md#locks

**Review by smallinsky — APPROVED** [review]
LGTM.

File: lib/auth/auth.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/auth/auth.go:1524]: It is a pointer https://github.com/gravitational/teleport/blob/018199304796348139968e66b9719f0051035166/lib/auth/auth.go#L1355 but I see that validation in done in https://github.com/gravitational/teleport/blob/018199304796348139968e66b9719f0051035166/lib/auth/auth.go#L1383. 
so my comment is not longer valid.

**Pull request merged**